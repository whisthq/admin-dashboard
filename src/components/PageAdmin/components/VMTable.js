import React, { Component } from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import ToggleButton from 'react-toggle-button'
import {
    faCircleNotch,
    faPlay,
    faPause,
} from '@fortawesome/free-solid-svg-icons'

import {
    startVM,
    deallocateVM,
    updateDB,
    setDev,
} from '../../../actions/index.js'

import moment from 'moment'
import { Table } from 'antd'
import 'antd/dist/antd.css'
import Style from '../../../styles/components/pageAdmin.module.css'

import '../../../static/App.css'

class VMTable extends Component {
    // intervalID var to keep track of auto-refreshing across functions
    intervalID

    componentDidMount() {
        this.props.dispatch(updateDB(false))

        // refresh the VM table every 60 seconds
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    getUpdatedDatabase() {
        this.props.dispatch(updateDB(false))
    }

    startVM = (vm_name) => {
        this.props.dispatch(startVM(vm_name))
    }

    deallocateVM = (vm_name) => {
        this.props.dispatch(deallocateVM(vm_name))
    }

    toggleDev = (mode, vm_name) => {
        this.props.dispatch(setDev(vm_name, !mode))
    }

    render() {
        let columns = []
        let data = []
        let component = this
        if (this.props.vm_info && this.props.vm_info.length) {
            Object.keys(this.props.vm_info[0]).forEach(function (key) {
                let fixWidth = false
                if (key === 'username') {
                    fixWidth = 200
                } else if (key === 'disk_name') {
                    fixWidth = 400
                } else if (
                    key === 'temporary_lock' ||
                    key === 'last_updated' ||
                    key === 'state'
                ) {
                    fixWidth = 150
                }
                let customRender = false
                if (key === 'temporary_lock') {
                    customRender = (unix) => (
                        <span> {moment(unix * 1000).format('lll')}</span>
                    )
                } else if (key === 'lock') {
                    customRender = (val) => <span>{val.toString()}</span>
                }
                if (key !== 'dev') {
                    columns.push({
                        title: key,
                        dataIndex: key,
                        sorter: (a, b) => {
                            if (a[key] === null) {
                                return 1
                            }
                            if (b[key] === null) {
                                return -1
                            }

                            var a_temp = a[key].toString().toLowerCase()
                            var b_temp = b[key].toString().toLowerCase()
                            if (a_temp > b_temp) {
                                return 1
                            } else if (a_temp < b_temp) {
                                return -1
                            }

                            return 0
                        },
                        width: fixWidth,
                        render: customRender,
                    })
                }
            })
            columns.push({
                title: 'dev',
                dataIndex: 'dev',
                render: (text, record, index) => (
                    <ToggleButton
                        value={record['dev']}
                        onToggle={(mode) => {
                            component.toggleDev(mode, record['vm_name'])
                        }}
                        colors={{
                            active: {
                                base: '#5EC4EB',
                            },
                            inactive: {
                                base: '#161936',
                            },
                        }}
                    />
                ),
                sorter: (a, b) => {
                    if (a['dev'] === null) {
                        return 1
                    }
                    if (b['dev'] === null) {
                        return -1
                    }
                    if (a['dev'] && !b['dev']) {
                        return 1
                    } else if (!a['dev'] && b['dev']) {
                        return -1
                    }
                    return 0
                },
            })
            columns.push({
                title: '',
                dataIndex: 'vmBtn',
                render: (text, record, index) =>
                    vmButton(
                        record['state'],
                        record['vm_name'],
                        record['lock']
                    ),
                width: 70,
            })
            this.props.vm_info.forEach(function (vm) {
                data.push(vm)
            })
        }
        columns.reverse()

        const vmButton = (state, vm_name, lock) => {
            const intermediate_states = [
                'DEALLOCATING',
                'STARTING',
                'ATTACHING',
                'STOPPING',
            ]
            if (
                this.props.vms_updating.includes(vm_name) &&
                intermediate_states.includes(state)
            ) {
                return (
                    <td
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faCircleNotch}
                            spin
                            style={{ color: '#111111' }}
                        />
                    </td>
                )
            } else if (
                (state === 'STOPPED' || state === 'DEALLOCATED') &&
                !lock
            ) {
                return (
                    <td
                        onClick={() => this.startVM(vm_name)}
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faPlay}
                            style={{ color: '#961418' }}
                        />
                    </td>
                )
            } else if (state === 'RUNNING_AVAILABLE' && !lock) {
                return (
                    <td
                        onClick={() => this.deallocateVM(vm_name)}
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faPause}
                            style={{ color: '#1ac447' }}
                        />
                    </td>
                )
            }
        }

        return (
            <Table
                columns={columns}
                dataSource={data}
                scroll={{ y: 400, x: 2000 }}
                size="middle"
                rowClassName={(record, index) =>
                    [
                        record['dev']
                            ? Style.blueBg
                            : record['lock'] ||
                              Number(record['temporary_lock']) >
                                  Math.round(new Date().getTime() / 1000)
                            ? Style.redBg
                            : Style.greenBg,
                        Style.tableRow,
                    ].join(' ')
                }
                onRow={(record, rowIndex) => {
                    return {
                        onClick: (event) => {
                            console.log(record)
                        },
                    }
                }}
                loading={!this.props.vmsUpdated}
            />
        )
    }
}

function mapStateToProps(state) {
    return {
        vm_info: state.AccountReducer.vm_info
            ? state.AccountReducer.vm_info.reverse()
            : [],
        vmsUpdated: state.AccountReducer.vmsUpdated,
        vms_updating: state.AccountReducer.vms_updating
            ? state.AccountReducer.vms_updating
            : [],
    }
}

export default connect(mapStateToProps)(VMTable)
