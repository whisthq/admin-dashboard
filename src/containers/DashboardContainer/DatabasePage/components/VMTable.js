import React, { Component } from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import moment from 'moment'
import { Table } from 'antd'

import '../../../../static/App.css'
import 'antd/dist/antd.css'
import Style from '../../../../styles/components/pageAdmin.module.css'

import {
    faCircleNotch,
    faPlay,
    faPause,
} from '@fortawesome/free-solid-svg-icons'

import { startVM, deallocateVM, fetchVMs } from '../../../../actions/index'

class VMTable extends Component {
    constructor(props) {
        super(props)

        this.state = {
            filterKeyword: null,
        }
    }

    // intervalID var to keep track of auto-refreshing across functions
    intervalID

    componentDidMount() {
        this.props.dispatch(fetchVMs(false))

        // refresh the VM table every 60 seconds
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    getUpdatedDatabase() {
        this.props.dispatch(fetchVMs(false))
    }

    startVM = (vm_name) => {
        this.props.dispatch(startVM(vm_name))
    }

    deallocateVM = (vm_name) => {
        this.props.dispatch(deallocateVM(vm_name))
    }

    keywordFilter = (e) => {
        let filterKeyword = e.target.value
        this.setState({
            filterKeyword: filterKeyword,
        })
    }

    render() {
        let columns = []
        let data = []
        let keys = [
            'vm_id',
            'user_id',
            'state',
            'location',
            'lock',
            'temporary_lock',
            'ip',
            'os',
        ]
        let component = this

        if (this.props.vm_info && this.props.vm_info.length) {
            columns.push({
                title: '',
                dataIndex: 'vmBtn',
                render: (text, record, index) =>
                    vmButton(record['state'], record['vm_id'], record['lock']),
                width: 70,
            })

            keys.forEach(function (key) {
                let fixWidth = false
                if (
                    key === 'temporary_lock' ||
                    key === 'location' ||
                    key === 'state' ||
                    key === 'vm_id' ||
                    key === 'ip'
                ) {
                    fixWidth = 150
                }
                let customRender = false
                if (key === 'temporary_lock') {
                    customRender = (val) => (
                        <span> {moment(val * 1000).format('lll')}</span>
                    )
                } else if (key === 'lock') {
                    customRender = (val) => <span>{val.toString()}</span>
                }
                columns.push({
                    title: key,
                    dataIndex: key,
                    sorter: (a, b) => {
                        if (a[key] === null) {
                            return 1
                        }
                        if (b[key] === null) {
                            return -1
                        }

                        var a_temp = a[key].toString().toLowerCase()
                        var b_temp = b[key].toString().toLowerCase()
                        if (a_temp > b_temp) {
                            return 1
                        } else if (a_temp < b_temp) {
                            return -1
                        }

                        return 0
                    },
                    width: fixWidth,
                    render: customRender,
                })
            })

            this.props.vm_info.forEach(function (vm) {
                data.push(vm)
            })
            if (component.state.filterKeyword) {
                data = data.filter((element) => {
                    return (
                        (element.vm_id &&
                            element.vm_id.includes(
                                component.state.filterKeyword
                            )) ||
                        (element.location &&
                            element.location.includes(
                                component.state.filterKeyword
                            )) ||
                        (element.ip &&
                            element.ip.includes(component.state.filterKeyword))
                    )
                })
            }
        }

        let vmButton = (state, vm_name, lock) => {
            const intermediate_states = [
                'DEALLOCATING',
                'STARTING',
                'ATTACHING',
                'STOPPING',
            ]
            if (
                this.props.vms_updating.includes(vm_name) &&
                intermediate_states.includes(state)
            ) {
                return (
                    <td
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faCircleNotch}
                            spin
                            style={{ color: '#111111' }}
                        />
                    </td>
                )
            } else if (
                (state === 'STOPPED' || state === 'DEALLOCATED') &&
                !lock
            ) {
                return (
                    <td
                        onClick={() => this.startVM(vm_name)}
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faPlay}
                            style={{ color: '#961418' }}
                        />
                    </td>
                )
            } else if (state === 'RUNNING_AVAILABLE' && !lock) {
                return (
                    <td
                        onClick={() => this.deallocateVM(vm_name)}
                        style={{
                            paddingLeft: 20,
                            paddingRight: 30,
                            fontSize: 11,
                        }}
                        className="pointerOnHover"
                    >
                        <FontAwesomeIcon
                            icon={faPause}
                            style={{ color: '#1ac447' }}
                        />
                    </td>
                )
            }
        }

        return (
            <div className={Style.tableWrapper}>
                <input
                    type="text"
                    placeholder="Filter by keyword"
                    style={{
                        width: '100%',
                        border: 'none',
                        padding: '10px 20px',
                        background: '#ebecf0',
                    }}
                    onChange={this.keywordFilter}
                />
                <Table
                    columns={columns}
                    dataSource={data}
                    scroll={{ y: 450, x: 1200 }}
                    size="middle"
                    rowClassName={(record, index) =>
                        [
                            record['lock'] ||
                            Number(record['temporary_lock']) >
                                Math.round(new Date().getTime() / 1000)
                                ? Style.redBg
                                : Style.greenBg,
                            Style.tableRow,
                        ].join(' ')
                    }
                    loading={!this.props.vmsUpdated}
                />
            </div>
        )
    }
}

function mapStateToProps(state) {
    return {
        vm_info: state.AccountReducer.vm_info
            ? state.AccountReducer.vm_info
            : [],
        vmsUpdated: state.AccountReducer.vmsUpdated,
        vms_updating: state.AccountReducer.vms_updating
            ? state.AccountReducer.vms_updating
            : [],
    }
}

export default connect(mapStateToProps)(VMTable)
