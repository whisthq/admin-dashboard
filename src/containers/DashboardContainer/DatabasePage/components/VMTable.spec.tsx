import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { Table } from 'antd'
import toJson from 'enzyme-to-json'

import { VMTable } from './VMTable'

Enzyme.configure({ adapter: new Adapter() })

/*
The VMTable is an ant-design table with custom buttons that have our own necessary functionality. It must display
a table and be able to fetch VM data and dispatch changes that need to be made.

Next time the code is updated please make it more legible.
*/
describe('<VMTable />', () => {
    // snapshot test
    it('renders correctly', () => {
        const wrapper = shallow(
            <VMTable
                // buncha dummy data
                vmsUpdated={false}
                vms_updating={[]}
                vms_info={[]}
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )

        expect(toJson(wrapper)).toMatchSnapshot()
    })

    test('vm_button helper is functional', () => {
        // TODO not possible to test until it is brought outside of the render() function
        // right now it's declared inside which is irritating and probably not optimal for performance
    })

    test('component mounts correctly with needed fetches and page change', () => {
        // TODO
    })
    test('component unmounts correctly', () => {
        // TODO
    })
    test('helpers work', () => {
        // TODO (this is optional)
        // this is less spec and more implementation, but you may want to use this test
        // or turn it into various tests to check whether your helper functions work
    })

    it('displays a table', () => {
        const wrapper = shallow(
            <VMTable
                // buncha dummy data
                vmsUpdated={false}
                vms_updating={[]}
                vms_info={[]}
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )

        expect(wrapper.find(Table)).toHaveLength(1)
    })
})
