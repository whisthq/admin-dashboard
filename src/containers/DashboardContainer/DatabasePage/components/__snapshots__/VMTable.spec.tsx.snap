// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VMTable /> renders correctly 1`] = `
<div
  className="tableWrapper"
>
  <input
    onChange={[Function]}
    placeholder="Filter by keyword"
    style={
      Object {
        "background": "#ebecf0",
        "border": "none",
        "padding": "10px 20px",
        "width": "100%",
      }
    }
    type="text"
  />
  <Table
    columns={Array []}
    dataSource={Array []}
    loading={true}
    rowClassName={[Function]}
    rowKey="key"
    scroll={
      Object {
        "x": 1200,
        "y": 450,
      }
    }
    size="middle"
  />
</div>
`;
